rl4co/models/common/constructive/base.py

```diff
@@ -0,0 +1,239 @@
+    ) -> Tuple[Tensor, Tensor]:
+        """Obtain heatmap logits for current action to the next ones
+        Args:
+            td: TensorDict containing the input data
+            hidden: Hidden state from the encoder. Can be any type
+            num_starts: Number of starts for multistart decoding
+
+        Returns:
+            Tuple containing the logits and the action mask
+        """
+        raise NotImplementedError("Implement me in subclass!")
+
+    def pre_decoder_hook(
+        self, td: TensorDict, env: RL4COEnvBase, hidden: Any = None, num_starts: int = 0
+    ) -> Tuple[TensorDict, Any, RL4COEnvBase]:
+        """By default, we don't need to do anything here.
+
+        Args:
+            td: TensorDict containing the input data
+            hidden: Hidden state from the encoder
+            env: Environment for decoding
+            num_starts: Number of starts for multistart decoding
+
+        Returns: # TODO
+            Tuple containing the updated hidden state, TensorDict, and environment
+        """
+        return td, env, hidden
+
+
+class NoEncoder(ConstructiveEncoder):
+    """Default encoder decoder-only models, i.e. autoregressive models that re-encode all the state at each decoding step."""
+
+    def forward(self, td: TensorDict) -> Tuple[Tensor, Tensor]:
+        """Return Nones for the hidden state and initial embeddings"""
+        return None, None
+
+
+class ConstructivePolicy(nn.Module):
+    """
+    TODO docstring
+
+        Note:
+            There are major differences between this decoding and most RL problems. The most important one is
+            that reward is not defined for partial solutions, hence we have to wait for the environment to reach a terminal
+            state before we can compute the reward with `env.get_reward()`.
+
+        Warning:
+        We suppose environments in the `done` state are still available for sampling. This is because in NCO we need to
+        wait for all the environments to reach a terminal state before we can stop the decoding process. This is in
+        contrast with the TorchRL framework (at the moment) where the `env.rollout` function automatically resets.
+        You may follow tighter integration with TorchRL here: https://github.com/ai4co/rl4co/issues/72.
+
+    """
+
+    def __init__(
+        self,
+        encoder: Union[ConstructiveEncoder, Callable],
+        decoder: Union[ConstructiveDecoder, Callable],
+        env_name: Union[str, RL4COEnvBase] = "tsp",
+        temperature: float = 1.0,
+        tanh_clipping: float = 0,
+        mask_logits: bool = True,
+        train_decode_type: str = "sampling",
+        val_decode_type: str = "greedy",
+        test_decode_type: str = "greedy",
+        **unused_kw,
+    ):
+        super(ConstructivePolicy, self).__init__()
+
+        if len(unused_kw) > 0:
+            log.error(f"Found {len(unused_kw)} unused kwargs: {unused_kw}")
```

Comment:
[Minor] should this be `warning` or `error`?

Reply:
Both are fine. I put `error` because passing non-existings kwargs should be discouraged and users should see that.
Btw, I think we should try to avoid `unused_kwargs` and similar for clarity